File: 3490-find-the-maximum-length-of-valid-subsequence-i/find-the-maximum-length-of-valid-subsequence-i.java
class Solution {
    public int maximumLength(int[] nums) {
        int n = nums.length;
        if (n == 2) return 2;

        // z: parity of first number (true for odd, false for even)
        boolean z = (nums[0] & 1) == 1;

        // len[0]: length of even parity subsequence
        // len[1]: length of odd parity subsequence
        // len[2]: length of alternating parity subsequence (starting with nums[0])
        int[] len = new int[3];
        len[z ? 1 : 0] = 1; // initialize correct parity count
        len[2] = 1;         // initialize alternating parity count

        for (int i = 1; i < n; i++) {
            boolean x = (nums[i] & 1) == 1; // parity of current number

            // Increase the count of current parity
            len[x ? 1 : 0]++;

            // Check if we can extend the alternating parity sequence
            if (x != z) {
                len[2]++;
                z = !z; // flip parity for alternating pattern
            }
        }

        // Return the maximum among even-only, odd-only, and alternating parity subsequences
        return Math.max(len[0], Math.max(len[1], len[2]));
    }
}

File: 3490-find-the-maximum-length-of-valid-subsequence-i/find-the-maximum-length-of-valid-subsequence-i.py
class Solution:
    def maximumLength(self, nums: List[int]) -> int:
        n = len(nums)

        if n == 2:
            return 2

        z = nums[0] % 2 == 1  # parity of first number

        length = [0, 0, 0]
        length[1 if z else 0] = 1
        length[2] = 1

        for i in range(1, n):
            x = nums[i] % 2 == 1  # parity of current number

            length[1 if x else 0] += 1

            if x != z:
                length[2] += 1
                z = not z

        return max(length[0], length[1], length[2])
